ic-payments/src/token_terminal.rs

```diff
@@ -0,0 +1,455 @@
+use std::sync::atomic::AtomicU64;
+
+use async_recursion::async_recursion;
+use candid::Principal;
+use ic_exports::ic_base_types::PrincipalId;
+use ic_exports::ic_icrc1::endpoints::TransferError;
+use ic_exports::ic_icrc1::{Account, Subaccount};
+use ic_exports::ic_kit::ic;
+use ic_helpers::tokens::Tokens128;
+
+use crate::error::{InternalPaymentError, PaymentError, RecoveryDetails, TransferFailReason};
+use crate::icrc1::{self, get_icrc1_balance, TokenTransferInfo};
+use crate::recovery_list::{RecoveryList, StableRecoveryList};
+use crate::transfer::{Operation, Stage, Transfer, TransferType};
+use crate::{Balances, TokenConfiguration, TxId};
+
+/// Id that is used by the terminal to specify that the transaction ID is unknown, but it knows for
+/// sure that the transaction exists.
+pub const UNKNOWN_TX_ID: u128 = u64::MAX as u128;
+
+/// Default number of retries in case of IC error, before a transfer stored into the list for
+/// recovery.
+const N_RETRIES: usize = 3;
+
+/// Default period when deduplication of a transaction is possible. This is set by the token
+/// implementation. 24 hours used here is the most common value, used by ICP and SNS-1 ledgers.
+const DEFAULT_DEDUP_PERIOD: u64 = 10u64.pow(9) * 60 * 60 * 24;
```

Comment:
Don't let @hagsteel see this. Since it's time related value, I recommend we use `Duration`, same with `TX_WINDOW`, more clear.

I think we can call `..as_nanos()` to be used in the `deduplication_period`
```suggestion
const DEFAULT_DEDUP_PERIOD: Duration = Duration::from_secs(60 * 60 * 24);
```

Reply:
Actually, I disagree with this position. In IC all time-related staff is u64, nowhere in there libs or internals they use Duration. So I'd rather stick to what the framework is using. @hagsteel

Reply:
Because API returns u64 and takes u64. Everyone who worked with IC would expect u64. Candid doesn't support Duration. Using Duration in this case just adds unnecessary complexity without solid benefit.